library/core/src/option.rs

```diff
@@ -2036,6 +2036,64 @@ impl<'a, T> const From<&'a mut Option<T>> for Option<&'a mut T> {
     }
 }
 
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<T> crate::marker::StructuralPartialEq for Option<T> {}
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<T: PartialEq> PartialEq for Option<T> {
+    #[inline]
+    default fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Some(l), Some(r)) => *l == *r,
+            (None, None) => true,
+            _ => false,
+        }
+    }
+}
+
+macro_rules! non_zero_option {
+    ( $( #[$stability: meta] $NZ:ty; )+ ) => {
+        $(
+            #[$stability]
+            impl PartialEq for Option<$NZ> {
+                #[inline]
+                fn eq(&self, other: &Self) -> bool {
+                    // SAFETY: Where T is core::num::NonZero*, Option<T> is
+                    // guarenteed to have the same size as T.
+                    unsafe {
+                        mem::transmute::<_, $NZ>(*self) ==
```

Comment:
This is UB, I think you mean to compare the wrapped integer.

Reply:
We should also update the safety comment to explicitly talk about the related invariant (i.e., validity of NonZero and the representation in Option) - transmute statically checks that the sizes match, that's not something that's really part of the unsafe bits we need to justify.